// rsuite/cjs/Tree/hooks/useTreeNodeProps.js
'use client';
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");
exports.__esModule = true;
exports.default = void 0;
var _react = _interopRequireWildcard(require("react"));
var _utils = require("../../utils");
var _hooks = require("../../internals/Picker/hooks");
var _TreeProvider = require("../TreeProvider");
var _Highlight = _interopRequireDefault(require("../../Highlight"));
function useTreeNodeProps(props) {
  var _useItemDataKeys = (0, _TreeProvider.useItemDataKeys)(),
    valueKey = _useItemDataKeys.valueKey,
    labelKey = _useItemDataKeys.labelKey,
    childrenKey = _useItemDataKeys.childrenKey;
  var _useCombobox = (0, _hooks.useCombobox)(),
    id = _useCombobox.id;
  var value = props.value,
    disabledItemValues = props.disabledItemValues,
    loadingNodeValues = props.loadingNodeValues,
    focusItemValue = props.focusItemValue,
    keyword = props.keyword,
    dragNode = props.dragNode,
    dragOverNodeKey = props.dragOverNodeKey,
    dropNodePosition = props.dropNodePosition;
  return (0, _react.useCallback)(function (nodeData, layer, index) {
    var DRAG_OVER = _utils.TREE_NODE_DROP_POSITION.DRAG_OVER,
      DRAG_OVER_TOP = _utils.TREE_NODE_DROP_POSITION.DRAG_OVER_TOP,
      DRAG_OVER_BOTTOM = _utils.TREE_NODE_DROP_POSITION.DRAG_OVER_BOTTOM;
    var visible = nodeData.visible;
    var draggingNode = dragNode !== null && dragNode !== void 0 ? dragNode : {};
    var nodeValue = nodeData[valueKey];
    var nodeLabel = nodeData[labelKey];
    var children = nodeData[childrenKey];
    var label = keyword ? /*#__PURE__*/_react.default.createElement(_Highlight.default, {
      as: "span",
      query: keyword
    }, nodeLabel) : nodeLabel;
    var dragging = (0, _utils.shallowEqual)(nodeValue, draggingNode[valueKey]);
    var dragStatus;
    if ((0, _utils.shallowEqual)(nodeValue, dragOverNodeKey)) {
      switch (dropNodePosition) {
        case DRAG_OVER:
          dragStatus = 'drag-over';
          break;
        case DRAG_OVER_TOP:
          dragStatus = 'drag-over-top';
          break;
        case DRAG_OVER_BOTTOM:
          dragStatus = 'drag-over-bottom';
          break;
      }
    }
    var disabled = disabledItemValues.some(function (disabledItem) {
      return (0, _utils.shallowEqual)(disabledItem, nodeValue);
    });
    var loading = loadingNodeValues.some(function (item) {
      return (0, _utils.shallowEqual)(item, nodeValue);
    });
    var active = (0, _utils.shallowEqual)(nodeValue, value);
    var focus = (0, _utils.shallowEqual)(nodeValue, focusItemValue);
    return {
      id: id ? id + "-opt-" + nodeValue : undefined,
      value: nodeValue,
      label: label,
      index: index,
      layer: layer,
      loading: loading,
      active: active,
      focus: focus,
      visible: visible,
      children: children,
      nodeData: nodeData,
      disabled: disabled,
      dragging: dragging,
      dragStatus: dragStatus
    };
  }, [childrenKey, disabledItemValues, dragNode, dragOverNodeKey, dropNodePosition, focusItemValue, id, keyword, labelKey, loadingNodeValues, value, valueKey]);
}
var _default = useTreeNodeProps;
exports.default = _default;
